File: minimum-absolute-difference-in-bst/minimum-absolute-difference-in-bst.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    
    //global variables to save state
    public TreeNode prev = null;
    public int minDiff = Integer.MAX_VALUE;

    public int getMinimumDifference(TreeNode root) {
        inorder(root);
        return minDiff;
    }

    //private helper method for recursion call stack
    private void inorder(TreeNode root){
        if (root == null) return;
        inorder(root.left); //traverse left

        // calculate current min
        if (prev != null) minDiff = minDiff < root.val - prev.val ? minDiff : root.val - prev.val;

        prev = root;
        inorder(root.right); //traverse right
    }
}

File: leetcode/Easy/minimum-absolute-difference-in-bst/minimum-absolute-difference-in-bst.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:

    #global variables to save state
    def __init__(self):
        self.prev = None
        self.minDiff = float('inf')

    def getMinimumDifference(self, root: Optional[TreeNode]) -> int:
        self.inorder(root)
        return self.minDiff
    
    #helper method for recursion
    def inorder(self,root):
        if not root: return

        self.inorder(root.left) # traversing left subtree

        #calcualting current difference
        if self.prev is not None: self.minDiff = self.minDiff if self.minDiff < root.val - self.prev.val else root.val - self.prev.val

        self.prev = root

        self.inorder(root.right) # traversing right subtree


        